tests/test_PR045_flatten_operation.py

```diff
@@ -0,0 +1,127 @@
+# BSD 3-Clause License; see https://github.com/jpivarski/awkward-1.0/blob/master/LICENSE
+
+import sys
+
+import pytest
+import numpy
+
+import awkward1
+
+def flatten(data, axis=0):
+    if axis < 0:
+        raise NotImplementedError("axis < 0 is much harder for untyped data...")
+    if isinstance(data, list):
+        if axis == 0:
+            if all(isinstance(x, list) for x in data):
+                return sum(data, [])
+            else:
+                raise ValueError("cannot concatenate non-lists")
+        else:
+            return [flatten(x, axis - 1) for x in data]
+    else:
+        raise ValueError("cannot flatten {0} objects".format(type(data)))
+
+def test_flatten_empty_array():
+    empty = awkward1.layout.EmptyArray()
+
+    assert awkward1.tolist(empty) == []
+    assert awkward1.tolist(empty.flatten()) == []
+
+def test_flatten_list_array():
+    content = awkward1.layout.NumpyArray(numpy.array([0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]))
+    starts  = awkward1.layout.Index64(numpy.array([0, 3, 5, 8]))
+    stops   = awkward1.layout.Index64(numpy.array([3, 3, 8, 9]))
+    array   = awkward1.layout.ListArray64(starts, stops, content)
+
+    assert awkward1.tolist(array) == [[0.0, 1.1, 2.2], [], [5.5, 6.6, 7.7], [8.8]]
+    assert awkward1.tolist(array.flatten()) == [0.0, 1.1, 2.2, 5.5, 6.6, 7.7, 8.8]
+
+    array2 = array[2:-1]
+    assert flatten(awkward1.tolist(array2)) == [5.5, 6.6, 7.7]
+    assert awkward1.tolist(array2.flatten()) == [5.5, 6.6, 7.7]
```

Comment:
Since this is only one subarray in `array[2:-1]`, it's not a great example. Maybe turn the `starts` and `stops` into:

```python
    starts  = awkward1.layout.Index64(numpy.array([0, 3, 3, 5, 8]))
    stops   = awkward1.layout.Index64(numpy.array([3, 3, 5, 8, 9]))
```

You still have unreachable elements, but `array[2:-1]` becomes `[[4.4, 5.5], [5.5, 6.6, 7.7]]`, which is non-trivial to flatten.